tianshou/env/venvs.py

```diff
@@ -1,21 +1,87 @@
+from tianshou.env.utils import gym_new_venv_step_type
 from tianshou.env.worker import (
     DummyEnvWorker,
     EnvWorker,
     RayEnvWorker,
     SubprocEnvWorker,
 )
 
+try:
+    import gym as old_gym
+    ENV_TYPE = Union[gym.Env, old_gym.Env]
+except ImportError:
+    old_gym = None
+    ENV_TYPE = gym.Env
+
 GYM_RESERVED_KEYS = [
     "metadata", "reward_range", "spec", "action_space", "observation_space"
 ]
 
 
+def _patch_env_generator(fn: Callable[[], ENV_TYPE]) -> Callable[[], gym.Env]:
+    """Takes an environment generator and patches it to return Gymnasium envs.
+
+    This function takes the environment generator `fn` and returns a patched
+    generator, without invoking `fn`. The original generator may return
+    Gymnasium or OpenAI Gym environments, but the patched generator wraps
+    the result of `fn` in a shimmy wrapper to convert it to Gymnasium,
+    if necessary.
+    """
+
+    def patched() -> Callable[[], gym.Env]:
+        assert callable(
+            fn
+        ), "Env generators that are provided to vector environemnts must be callable."
+        env = fn()
+        if isinstance(env, gym.Env):
+            return env
+
+        if old_gym is None or not isinstance(env, old_gym.Env):
+            raise ValueError(
+                f"Environment generator returned a {type(env)}, not a Gymnasium "
+                f"environment. In this case, we expect OpenAI Gym to be "
+                f"installed and the environment to be an OpenAI Gym environment."
+            )
+        try:
+            import shimmy
+        except ImportError as e:
+            raise ImportError(
+                "Missing shimmy installation. You provided an environment generator "
+                "that returned an OpenAI Gym environment. "
+                "Tianshou has transitioned to using Gymnasium internally. "
+                "In order to use OpenAI Gym environments with tianshou, you need to "
+                "install shimmy (`pip install shimmy`)."
+            ) from e
+
+        warnings.warn(
+            "You provided an environment generator that returned an OpenAI Gym "
+            "environment. We strongly recommend transitioning to Gymnasium "
+            "environments. "
+            "Tianshou is automatically wrapping your environments in a compatibility "
+            "layer, which could potentially cause issues."
+        )
+
+        gym_version = packaging.version.parse(old_gym.__version__)
+        if gym_version >= packaging.version.parse("0.26.0"):
+            return shimmy.GymV26CompatibilityV0(env=env)
+        elif gym_version >= packaging.version.parse("0.22.0"):
+            return shimmy.GymV22CompatibilityV0(env=env)
+        else:
+            raise Exception(
+                f"Found OpenAI Gym version {gym.__version__}. "
+                f"Tianshou only supports OpenAI Gym environments of version>=0.22.0"
+            )
```

Comment:
@pseudo-rnd-thoughts Can you confirm that this is how the shimmy wrappers are intended to be used? I.e. for any environment `0.26.0>gym.__version__>=0.22.0` we use `GymV22CompatibilityV0"` and for `gym.__version__ >= 0.26.0` we use `GymV26CompatibilityV0`?